### ISSUE TITLE
Add custom User-Agent for Async Post requests

### ISSUE THREAD

**Issue by ovidiul** [woocommerce/action-scheduler#787]
When dispatching an [async](https://github.com/woocommerce/action-scheduler/blob/7f4aaf0e011bb0b41b266afd479f6935a495231b/lib/WP_Async_Request.php#L94) QueueRunner request, the User-Agent sent with that POST request usually defaults to the WordPress default value.

Right now, it's pretty hard to identify if the requests to the 'wp-admin/admin-ajax.php' endpoint came from Action Scheduler or another plugin that used similar request parameters.

Would it be worth adding a filter for the User-Agent as well as adjusting the default User-Agent sent by Action Scheduler async requests?

Happy to work on a PR for this.

**Comment by peterfabian** [comment]
Sounds like a good idea, thanks @ovidiul .

**Comment by barryhughes** [comment]
@peterfabian | @ovidiul 

Though I like this idea and *suspect* it is low risk, it does still feel like there is a chance of accidentally causing disruption in some cases (suppose there is a deployment out there where user agent strings for admin-ajax requests need to be on an approved list, or suppose internal admin-ajax requests are already being routed to a different box on the basis of their user agent string). 

Instead, we could flip this round and continue to implicitly use the default WP user agent string, while making it possible to customize things on a project-by-project or request-by-request basis. Two ways we might do this:

- **Low effort** → the existing PR [already takes care of making the user agent string filterable](https://github.com/woocommerce/action-scheduler/pull/789/commits/2c66e168d5e105c98f763926930e455ec242666f#diff-1bc85997145e00ddcc40255901b06d55830851841501e079682ab9f05d978eaeR144): we could tweak this and make it so this value is only added to the args array if it is set to something other than `null`.
- **Slightly more effort** → looking at the latest [upstream version of `WP_Async_Request`](https://github.com/deliciousbrains/wp-background-processing/blob/1.0.2/classes/wp-async-request.php#L154) the entire post args array (and many other things besides) are now filterable, whereas they weren't previously. We could take advantage of this.

If we took the second approach, we would likely also want to create a child class with its own custom identifier (and projects could then override the user agent string or other args by filtering on something like `as_async_request_post_args`).

Then again, perhaps I'm seeing risks that don't exist. What do you think?

**Comment by ovidiul** [comment]
Thanks for sharing the details @barryhughes 

I do agree that this change might come in with potential backward compatibility issues, so probably changing the User Agent might be unsafe.

The second option that implies more effort does make sense and I would vote for that, but do you see any issues that might appear from actually upgrading the original library instead?

Happy to help with testing that if needed.

**Comment by peterfabian** [comment]
I also agree it's a safer approach to leave the original value and make it changeable. Updating the `WP_Async_Request` sounds like the right choice.

**Comment by barryhughes** [comment]
> Do you see any issues that might appear from actually upgrading the original library instead?

We'd definitely need to do some testing to be confident, but it looks like the method signatures in the [updated abstract class](https://github.com/deliciousbrains/wp-background-processing/blob/1.0.2/classes/wp-async-request.php) are unchanged (the only real difference is the addition of a bunch of filters)—so no red flags that I can see so far.

It also turns out we already set a [custom prefix and custom action](https://github.com/woocommerce/action-scheduler/blob/3.4.0/classes/ActionScheduler_AsyncRequest_QueueRunner.php#L21-L35) in our existing concrete class so no changes are needed there.

If we make this change, we will probably also want to document those new hooks over on [actionscheduler.org](https://actionscheduler.org/). That is:

- `as_aync_request_queue_runner_post_args`
- `as_aync_request_queue_runner_query_args`
- `as_aync_request_queue_runner_query_url`

(Which we'd do after the release, assuming we go ahead with this approach: just noting it so we don't forget.)

**Comment by ovidiul** [comment]
Created the PR https://github.com/woocommerce/action-scheduler/pull/794 with the code update for that library, it does not show any significant changes other than the obvious `apply_filters` add-on.

**Comment by barryhughes** [comment]
Merged #794—likely will ship in 3.5.0.
